File: Emulator/src/tests/RAM.ts
import test from 'ava'
import { RAM } from '../components/RAM'

test('RAM can init', t => {
	t.not(new RAM(), null)
})

test('ROM can write', t => {
	const ram = new RAM()

	ram.write(0x0200, 0xAA)
	ram.write(0x0400, 0x55)

	t.is(ram.data[0x0200], 0xAA)
	t.is(ram.data[0x0400], 0x55)
})

test('ROM can read', t => {
	const ram = new RAM()

	t.is(ram.read(0x0000), 0x00)
	t.is(ram.read(0x7FFF), 0x00)
})

File: Firmware/DB Emulator/lib/DBEmulator/RAM/RAM.cpp
#include "RAM.h"

RAM::RAM() {
  uint16_t size = RAM_END - RAM_START + 1;

  this->data = new uint8_t[size];

  for (uint16_t i = 0x0000; i < size; i++) {
    this->data[i] = 0x00; 
  }
}

RAM::~RAM() {
  delete[] this->data;
}

uint8_t RAM::read(uint16_t index) {
  if (index <= (RAM_END - RAM_START)) {
    return this->data[index];
  } else {
    return 0x00;
  }
}

void RAM::write(uint16_t index, uint8_t value) {
  if (index <= (RAM_END - RAM_START)) {
    this->data[index] = value;
  }
}